[sc-68521] [WIP] Make `AmplitudeEmbedding` an alias of `StatePrep`

KetpuntoG opened the pull request:
This PR seeks to simplify `AmplitudeEmbedding` so that it simply references `StatePrep`. 
To do this, I have moved the `padding` and `normalize` logic to `StatePrep`.

KetpuntoG commented:
[sc-68521]

soranjh commented on pennylane/math/quantum.py:
```diff
@@ -19,7 +19,7 @@
 
 from autoray import numpy as np
-from numpy import float64
+from numpy import float64  # pylint:disable=wrong-import-order
```
We should have a better fix for this.

soranjh replied:
Please run `isort` on this file and remove disabling pylint. It will likely fix the issue. Here is the imports I locally created with `isort`:

```
import functools
# pylint: disable=import-outside-toplevel
import itertools
from string import ascii_letters as ABC

import pennylane as qml
from autoray import numpy as np

from numpy import float64

from . import single_dispatch  # pylint:disable=unused-import
from .matrix_manipulation import _permute_dense_matrix
from .multi_dispatch import (diag, dot, einsum, get_interface,
                             scatter_element_add)
from .utils import allclose, cast, cast_like, convert_like, is_abstract
```

KetpuntoG replied:
It does not work. It is in conflict with CodeFactor
There is [this](https://xanaduhq.slack.com/archives/CA89H1BAN/p1723647734543739) open discussion

soranjh commented on pennylane/ops/qubit/state_preparation.py:
```diff
@@ -151,6 +159,8 @@ class StatePrep(StatePrepBase):
     Args:
         state (array[complex]): a state vector of size 2**len(wires)
```
Whit the new padding, does it still need to be of size `2**len(wires)`?

soranjh commented on pennylane/ops/qubit/state_preparation.py:
```diff
@@ -127,9 +129,15 @@ def state_vector(self, wire_order: Optional[WiresLike] = None) -> TensorLike:
+    :math:`2^n` where :math:`n` is the number of qubits used in the embedding.
+
+    To represent a valid quantum state vector, the L2-norm of ``state`` must be one.
+    The argument ``normalize`` can be set to ``True`` to automatically normalize the features.
```
You mean normalize the state?

soranjh commented on pennylane/ops/qubit/state_preparation.py:
```diff
@@ -172,23 +182,25 @@ class StatePrep(StatePrepBase):
+    # pylint: disable=unused-argument
     @staticmethod
-    def compute_decomposition(state: TensorLike, wires: WiresLike) -> list[Operator]:
+    def compute_decomposition(state: TensorLike, wires: WiresLike, **kwargs) -> list[Operator]:
```
Why this addition is needed?

KetpuntoG replied:
because now in this template `normalize` and `pad_with` are also passed to the decomposition

soranjh commented on pennylane/templates/embeddings/amplitude.py:
```diff
@@ -38,11 +36,6 @@ class AmplitudeEmbedding(StatePrep):
 
         On some devices, ``AmplitudeEmbedding`` must be the first operation of a quantum circuit.
 
-    .. warning::
```
This is not valid anymore?

KetpuntoG replied:
Now it works the same as StatePrep and it is differentiable (although I tested AmplitudeEmbedding before the change  and also seemed to be differentiable. Maybe it was fixed and the warning was not deleted).

soranjh commented on tests/devices/test_default_qubit_jax.py:
```diff
@@ -438,7 +438,7 @@ def circuit(x):
             return qml.expval(qml.PauliZ(wires=0))
 
-        with pytest.raises(ValueError, match="Sum of amplitudes-squared does not equal one."):
+        with pytest.raises(ValueError, match="The state must be a vector of norm 1.0"):
```
Why this is changed?

KetpuntoG replied:
While checking the same thing in both templates, each of those returned a different message. When unifying them I had to choose one. As you can see the meaning is equivalent

soranjh commented on tests/devices/test_default_qubit_tf.py:
```diff
@@ -352,7 +352,7 @@ def test_invalid_state_prep_size(self):
         state = np.array([0, 1])
 
-        with pytest.raises(ValueError, match=r"State vector must have shape \(2\*\*wires,\)"):
+        with pytest.raises(ValueError, match=r"State must be of length 4"):
```
Why this is changed?

KetpuntoG replied:
Same as before. The error occurs for the same reason with both templates but I had to choose one of the two default messages when unifying the templates.

soranjh commented on tests/tape/test_tape.py:
```diff
@@ -804,11 +804,11 @@ def test_array_parameter(self):
-        assert tape.get_parameters() == params
 
-        b = np.array([0, 1, 0, 0])
+        b = np.array([0.0, 1.0, 0.0, 0.0])
```
Why this is changed?

KetpuntoG replied:
The problem is that now I preprocess in StatePrep and it is transforming it by default into a float array
If I leave it as integers, it gives me problems in differentiation 🤔

soranjh commented on tests/templates/test_embeddings/test_amplitude.py:
```diff
@@ -199,7 +200,10 @@ def circuit(x=None):
             qml.AmplitudeEmbedding(features=x, wires=range(n_qubits))
             return qml.expval(qml.PauliZ(0))
 
-        with pytest.raises(ValueError, match="Features must be a one-dimensional (tensor|vector)"):
+        with pytest.raises(
+            ValueError,
+            match="State must be a one-dimensional tensor, or two-dimensional with batching;",
+        ):
```
Why this is changed?

KetpuntoG replied:
unification of error messages

soranjh commented on tests/templates/test_subroutines/test_qubitization.py:
```diff
@@ -138,30 +138,42 @@ def test_legacy_new_opmath():
-                qml.adjoint(qml.AmplitudeEmbedding(np.array([1.0, 1.0]) / np.sqrt(2), wires=[1])),
+                qml.adjoint(
+                    qml.AmplitudeEmbedding(
+                        np.array([1.0, 1.0]) / np.sqrt(2), wires=[1], pad_with=0, normalize=True
```
Why the new args are needed now?

KetpuntoG replied:
It seems that there was a small bug in `AmplitudeEmbedding` because it was not considering `normalize` and `pad_with` as hyperparameters when they are. When correcting this, this test complained because it had to check that `pad_with` and `normalize` also matched in the decomposition

soranjh reviewed (changes requested):
The PR looks good, but there are many changes in the tests, that I am not sure why they are needed.

KetpuntoG commented:
> The PR looks good, but there are many changes in the tests, that I am not sure why they are needed.

Thanks @soranjh ! I answered all the questions and addressed the comments. 
(Still pending [this one](https://github.com/PennyLaneAI/pennylane/pull/6034#discussion_r1720229983))

obliviateandsurrender commented on pennylane/ops/qubit/state_preparation.py:
```diff
@@ -127,9 +129,15 @@ def state_vector(self, wire_order: Optional[WiresLike] = None) -> TensorLike:
 
 class StatePrep(StatePrepBase):
-    r"""StatePrep(state, wires)
+    r"""StatePrep(state, wires, pad_with, nomralize)
```
```suggestion
    r"""StatePrep(state, wires, pad_with, normalize)
```

This also misses out `default` values in the documentation.

obliviateandsurrender commented on pennylane/ops/qubit/state_preparation.py:
```diff
@@ -127,9 +129,15 @@ def state_vector(self, wire_order: Optional[WiresLike] = None) -> TensorLike:
+    r"""StatePrep(state, wires, pad_with, nomralize)
     Prepare subsystems using the given ket vector in the computational basis.
 
+    By setting ``pad_with`` to a real or complex number, ``state`` is automatically padded to dimension
```
What does this mean? 🤔 What is the `embedding` you mention here?

KetpuntoG replied:
Changed to "template"

obliviateandsurrender commented on pennylane/ops/qubit/state_preparation.py:
```diff
@@ -235,6 +259,73 @@ def state_vector(self, wire_order: Optional[WiresLike] = None):
+        norm = math.linalg.norm(state, axis=-1)
+
+        if math.is_abstract(norm):
+            if normalize or pad_with:
```
Why can we not simply set `normalize=True` when `pad_with` is not `None`?

obliviateandsurrender commented on pennylane/ops/qubit/state_preparation.py:
```diff
@@ -235,6 +259,73 @@ def state_vector(self, wire_order: Optional[WiresLike] = None):
+        # normalize
+        if "int" in str(state.dtype):
+            state = math.cast_like(state, 0.0)
+        norm = math.linalg.norm(state, axis=-1)
```
I think computing `norm` is the computation bottleneck and not renormalization via division. This makes me wonder, how does `normalization` keyword actually helps lol.

KetpuntoG replied:
I see your point, I can compute the norm only if `normalize = True`.
Should I write `math.is_abstract(state)` then?

obliviateandsurrender replied:
That would be ideal. Yes, I think you can do the renormalization when the `state` is `abstract` and when `normalize=True`.

KetpuntoG replied:
Oh but now that I see it, then you would be suggesting to remove the error message that an input is not normalized

obliviateandsurrender replied:
Please add a `TODO` for this as well.

obliviateandsurrender commented on tests/tape/test_tape.py:
```diff
@@ -806,7 +806,7 @@ def test_array_parameter(self):
         assert tape.get_parameters() == params
 
-        b = np.array([0, 1, 0, 0])
+        b = np.array([0.0, 1.0, 0.0, 0.0])
```
What happens when one uses vectors with `dtype=int`? 🤔

KetpuntoG replied:
I think I'll fix it with this last commit

The problem is that now I preprocess in StatePrep and it is transforming it by default into a float array

KetpuntoG replied:
Hmm didn't fix it, doesn't seem like an easy thing to avoid

obliviateandsurrender commented on pennylane/ops/qubit/state_preparation.py:
```diff
@@ -211,6 +223,18 @@ def compute_decomposition(state: TensorLike, wires: WiresLike) -> list[Operator]
         """
         return [MottonenStatePreparation(state, wires)]
```
We should mention in the docstring that we are using `Mottonen` for the the decomposition on the un-supported devices.

obliviateandsurrender reviewed (changes requested):
Looking good @KetpuntoG! Leaving some queries and suggestions.

KetpuntoG commented on pennylane/ops/qubit/state_preparation.py:
```diff
@@ -28,6 +28,8 @@
 
 state_prep_ops = {"BasisState", "StatePrep", "QubitDensityMatrix"}
 
+TOLERANCE = 1e-10
```
Currently is global in [AmplitudeEmbedding](https://github.com/PennyLaneAI/pennylane/blob/d7509cddf63a110fe4a22dd6914a12251aca722d/pennylane/templates/embeddings/amplitude.py#L23)

obliviateandsurrender replied:
We should aim to remove it in the future. Might want to add a `TODO` for this.

obliviateandsurrender commented on pennylane/ops/qubit/state_preparation.py:
```diff
@@ -211,6 +223,18 @@ def compute_decomposition(state: TensorLike, wires: WiresLike) -> list[Operator]
         return [MottonenStatePreparation(state, wires)]
 
+    def _flatten(self):
+        print(self.hyperparameters)
```
```suggestion
```

obliviateandsurrender commented on pennylane/ops/qubit/state_preparation.py:
```diff
@@ -149,10 +158,12 @@ class StatePrep(StatePrepBase):
-        state (array[complex]): a state vector of size 2**len(wires)
+        state (array[complex]): the state vector to prepare
         wires (Sequence[int] or int): the wire(s) the operation acts on
+        pad_with (float or complex): the value to pad the state vector with
```
```suggestion
        pad_with (float | complex | None): the value to pad the state vector with
```

soranjh replied:
Please do not upgrade the UI in this PR, we can have a follow-up PR for all these upgrades, that should be approved by Product.

soranjh replied:
Also, why the original description from AmplitudeEmbedding is changed? It states:

pad_with (float or complex): if not None, the input is padded with this constant to size :math:`2^n`

KetpuntoG replied:
updated 👍

obliviateandsurrender commented on pennylane/ops/qubit/state_preparation.py:
```diff
@@ -172,23 +183,25 @@ class StatePrep(StatePrepBase):
+        self,
+        state: TensorLike,
+        wires: WiresLike,
+        pad_with=None,
+        normalize=False,
```
```suggestion
        pad_with: Optional[complex] = None,
        normalize: Optional[bool] = False,
```

soranjh replied:
Please do not upgrade the UI in this PR, we can have a follow-up PR for all these upgrades, that should be approved by Product.

obliviateandsurrender commented on pennylane/ops/qubit/state_preparation.py:
```diff
@@ -235,6 +259,76 @@ def state_vector(self, wire_order: Optional[WiresLike] = None):
+
+        # normalize
+        if "int" in str(state.dtype):
+            state = math.cast_like(state, 0.0)
```
What are we doing here?

soranjh replied:
And why it was not needed in the original AmplitudeEmbedding?

KetpuntoG replied:
It was indeed in AmplitudeEmbedding [[link](https://github.com/PennyLaneAI/pennylane/blob/dea7a2d0ac4e9ce70b416b143a54072d45933ba5/pennylane/templates/embeddings/amplitude.py#L202)]

The issue is due to the fact that `torch` and `tensorflow` fail when calculating the norm of an integer vector. In the past this was not a problem for StatePrep but now that we do preprocessing, this is visible. So that the error is not raised, the casting is done.



<img width="785" alt="Captura de pantalla 2024-08-20 a las 20 21 12" src="https://github.com/user-attachments/assets/b68aeb8b-f173-4eb0-b48d-5f1032f950f6">

soranjh replied:
OK, it has been added recently [here](https://github.com/PennyLaneAI/pennylane/pull/5969). Thanks.

obliviateandsurrender commented on tests/ops/qubit/test_state_prep.py:
```diff
@@ -82,6 +82,32 @@ def test_StatePrep_decomposition(self):
         assert isinstance(ops1[0], qml.MottonenStatePreparation)
         assert isinstance(ops2[0], qml.MottonenStatePreparation)
 
+    def test_StatePrep_padding(self):
```
follow snake case naming here and below - `test_state_prep_padding`

KetpuntoG replied:
Are you sure here? I'm writing StatePrep referring to the template

obliviateandsurrender commented on tests/ops/qubit/test_state_prep.py:
```diff
@@ -82,6 +82,32 @@ def test_StatePrep_decomposition(self):
+
+        @qml.qnode(qml.device("default.qubit", wires=2))
+        def circuit():
+            qml.StatePrep(state, pad_with=0, wires=wires)
```
Can we parameterize this test with both float and complex padding values as well.

soranjh replied:
Is this comment addressed?

obliviateandsurrender commented on tests/ops/qubit/test_state_prep.py:
```diff
@@ -82,6 +82,32 @@ def test_StatePrep_decomposition(self):
+    def test_StatePrep_normalize(self):
+        """Test that StatePrep normalizes the input state correctly."""
+
+        state = np.array([1, 1, 1, 1])
```
We should test this for at least one complex state.

soranjh commented on tests/ops/qubit/test_state_prep.py:
```diff
@@ -82,6 +82,45 @@ def test_StatePrep_decomposition(self):
+    def test_StatePrep_padding(self, state, expected):
+        """Test that StatePrep pads the input state correctly."""
+
+        print(state)
```
```suggestion
```

soranjh reviewed (approved):
Looks good to me, please address my final two comments before merging.

obliviateandsurrender commented on tests/tape/test_tape.py:
```diff
@@ -1007,7 +1007,9 @@ def test_depth_expansion(self):
-                qml.StatePrep([0, 1, 0, 0], wires=[0, 1]),  # still a StatePrepBase :/
+                qml.MottonenStatePreparation(
+                    [0, 1, 0, 0], wires=[0, 1]
+                ),  # still a StatePrepBase :/
```
do we still need this comment?

obliviateandsurrender reviewed (approved):
Thanks @KetpuntoG! Overall, looks good to me.
